integration-tests/narayana-lra/src/main/java/org/acme/quickstart/lra/coordinator/TransactionalResource.java

```diff
@@ -0,0 +1,98 @@
+package org.acme.quickstart.lra.coordinator;
+
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_CONTEXT_HEADER;
+
+import java.net.URI;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import org.eclipse.microprofile.lra.annotation.Compensate;
+import org.eclipse.microprofile.lra.annotation.Complete;
+import org.eclipse.microprofile.lra.annotation.ws.rs.LRA;
+
+@Path("/tx")
+public class TransactionalResource {
+    static AtomicInteger completions = new AtomicInteger(0);
+    static AtomicInteger compensations = new AtomicInteger(0);
+
+    // run a method that starts an LRA on entry and does not end it on exit
+    @POST
+    @Path("/start")
+    @LRA(end = false)
+    public Response startTx(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return Response.created(lraId).entity(lraId).build();
+    }
+
+    // run a method must be called with an active LRA and ends the LRA on exit
+    @PUT
+    @Path("/end")
+    @LRA(LRA.Type.MANDATORY)
+    public String endTx(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return lraId.toASCIIString();
+    }
+
+    @Path("/status")
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String status() {
+        return "hello";
+    }
+
+    // run a method that starts an LRA on entry and ends it on exit
+    @POST
+    @Path("/lra")
+    @LRA
+    public Response doInTx(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return Response.created(lraId).entity(lraId).build();
+    }
+
+    // return the number of times the completion callback was called
+    @GET
+    @Path("completions")
+    public int completions() {
+        return completions.get();
+    }
+
+    // return the number of times the compensation callback was called
+    @GET
+    @Path("compensations")
+    public int compensations() {
+        return compensations.get();
+    }
+
+    // callback to inform the participant service that the LRA is cancelling
+    @PUT
+    @Path("compensate")
+    @Compensate
+    public Response compensateWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId, String userData) {
```

Comment:
That's a shame, users would find it extremely useful. I will check whether or not it still works in the narayana implementation of the LRA spec and if not then I will remove it.

Reply:
we can have this on top of the spec in the Narayana implementation as an experimental feature, agreed it would be nice to have.

Reply:
It's not an experimental feature, it is part of the narayana implementation of LRAs and has been present, I believe (I would need to look through the history) ever since I took the spec to the MP group. Don't forget that the MP-LRA spec process removed a lot of useful stuff from the implementation and  that the narayana LRA implementation is not restricted to what the spec says, it is a transactional solution that just happens to implement all of the requirements of the MP-LRA spec.

Reply:
true, not experimental, additional, not speced. Sorry.